commands/create_role_button.go

```diff
@@ -0,0 +1,110 @@
+	"github.com/disgoorg/json"
+	"github.com/myrkvi/heimdallr/utils"
+)
+
+var CreateRoleButtonCommand = discord.SlashCommandCreate{
+	Name: "create-role-button",
+	NameLocalizations: map[discord.Locale]string{
+		discord.LocaleNorwegian: "lag-rolleknapp",
+	},
+	Description: "Create a button that assigns a role to the user when clicked.",
+	DescriptionLocalizations: map[discord.Locale]string{
+		discord.LocaleNorwegian: "Lag ein knapp som gjev brukaren ei rolle når han vert trykt på.",
+	},
+
+	DMPermission:             utils.Ref(false),
+	DefaultMemberPermissions: json.NewNullablePtr(discord.PermissionManageRoles),
+
+	Options: []discord.ApplicationCommandOption{
+		discord.ApplicationCommandOptionRole{
+			Name: "role",
+			NameLocalizations: map[discord.Locale]string{
+				discord.LocaleNorwegian: "rolle",
+			},
+			Description: "The role to assign to the user.",
+			DescriptionLocalizations: map[discord.Locale]string{
+				discord.LocaleNorwegian: "Rollen som skal gjevast til brukaren.",
+			},
+			Required: true,
+		},
+		discord.ApplicationCommandOptionString{
+			Name: "instructions",
+			NameLocalizations: map[discord.Locale]string{
+				discord.LocaleNorwegian: "instruksjonar",
+			},
+			Description: "Instructions to display to the user above the button.",
+			DescriptionLocalizations: map[discord.Locale]string{
+				discord.LocaleNorwegian: "Instruksjonar som skal visast til brukaren over knappen.",
+			},
+			Required: false,
+		},
+		discord.ApplicationCommandOptionString{
+			Name: "text",
+			NameLocalizations: map[discord.Locale]string{
+				discord.LocaleNorwegian: "tekst",
+			},
+			Description: "The text to display on the button.",
+			DescriptionLocalizations: map[discord.Locale]string{
+				discord.LocaleNorwegian: "Teksten som skal visast på knappen.",
+			},
+			Required: false,
+		},
+	},
+}
+
+func CreateRoleButtonHandler(e *handler.CommandEvent) error {
+	if e.GuildID() == nil {
+		slog.Warn("Received create role button command in DMs or guild ID is otherwise nil")
+		return ErrEventNoGuildID
+	}
+
+	slog.Info(
+		"Received create role button command",
+		"guildID",
+		e.GuildID(),
+		"user",
+		e.User().ID,
+	)
+
+	permissions := e.Member().Permissions
+
+	role := e.SlashCommandInteractionData().Role("role")
+
+	instructions := e.SlashCommandInteractionData().String("instructions")
+	if instructions == "" {
+		instructions = fmt.Sprintf("Click the button below to get the **%s** role.", role.Name)
+	}
+
+	text := e.SlashCommandInteractionData().String("text")
+	if text == "" {
+		text = "Get role"
+	}
+
+	// Check if the user has permission to assign roles
+	if !permissions.Has(discord.PermissionManageRoles) {
+		_ = respondWithContentEph(e, "You don't have permissions to assign roles.")
+		return nil
+	}
+
+	// Check if the specific role in question is one the user can assign
+	if !permissions.Has(role.Permissions) {
+		_ = respondWithContentEph(e, "You don't have permissions to assign that role.")
+		return nil
+	}
```

Comment:
This could result in some weird things, in which someone can override the slash command requirement to be something that doesn't have this permission and leave them effectively unable to use the command. This *could* be the desired behaviour, in that we do not want anyone that does not have the permission to be able to use it, regardless, thus hardcoding a permission requirement.

If that's the case, maybe there should be a more informative error message regarding it.

Reply:
wouldn't that require them having the role with the negating permission be a higher priority up the list? in which case, we can check for that when the component is interacted with, right?

Reply:
I don't think I get entirely what you're saying, but as for roles and permissions, they cannot be negated on the server/role level. Channel overrides can negate permissions for people/roles, but you would then have to check each channel.

Reply:
I thought what you were trying to say originally is that someone getting a role may lock them out of a room they originally had access to. If all you're saying is we should be clear about that you need to have permissions for the role you're creating the button for, sure. This check prevents someone using this command to get permissions they didn't already have.

Reply:
Resolved in a241461

Reply:
The permission required to use a slash command is here set to be the default of requiring someone to have the ability to manage roles. This can be overridden, and a server admin can allow any person or role to access the command. Here, there is a hardcoded check that will not let the user use the command if they do not have the permission to manage roles, even if the user is allowed to use the command (i.e. if a server admin has added an override).

The second check was a little unclear to me in its intent in the first commit, but yeah that makes sense, I agree.

Reply:
Are you saying you _want_ people to override that requirement?

Reply:
Or is 0878d77 all you were getting at?

Reply:
![Discord view of overrides for the Approve context command shown on users](https://github.com/user-attachments/assets/cea63705-0b3d-4322-b05f-8cd4f77e932f)  
![Discord view of default permission required to use Approve context command](https://github.com/user-attachments/assets/5affa98d-4d76-49a7-85b5-3f4148ec1b34)

E.g. for Heimdallr's `Approve` user context command, it defaults to requiring users to have the *Kick Members* permission to use it. However, adding an override, any user or role added will be able to use it without having the *Kick Members* permission.

Slash commands will always let admins override who can "use" the slash command. We can enforce a hardcoded requirement despite this, but it should be reasoned why an administrator cannot allow someone without that permission to use the command.